components/abstractions/src/main/java/com/microsoft/kiota/store/InMemoryBackingStore.java

```diff
@@ -212,39 +232,50 @@ private void setupNestedSubscriptions(
         }
     }
 
-    private void ensureCollectionPropertyIsConsistent(final String key, final Object storeItem) {
-        if (storeItem instanceof Pair) { // check if we put in a collection annotated with the size
-            final Pair<?, Integer> collectionTuple = (Pair<?, Integer>) storeItem;
-            Object[] items;
-            if (collectionTuple.getValue0() instanceof Collection) {
-                items = ((Collection<Object>) collectionTuple.getValue0()).toArray();
-            } else { // it is a map
-                items = ((Map<?, Object>) collectionTuple.getValue0()).values().toArray();
-            }
+    private void ensureCollectionPropertiesAreConsistent() {
+        HashMap<String, Object> currentStoreDirtyCollections = new HashMap<>();
+        List<BackedModel> nestedBackedModelsToEnumerate = new ArrayList<>();
 
-            for (final Object item : items) {
-                touchNestedProperties(item); // call get on nested properties
+        for (final Map.Entry<String, Pair<Boolean, Object>> entry : this.store.entrySet()) {
+            final Pair<Boolean, Object> wrapper = entry.getValue();
+            if (wrapper.getValue1() instanceof Pair) {
+                final Pair<?, Integer> collectionTuple = (Pair<?, Integer>) wrapper.getValue1();
+                Object[] items = getObjectArrayFromCollectionWrapper(collectionTuple);
+                for (Object item : items) {
+                    if (!(item instanceof BackedModel)) break;
+                    nestedBackedModelsToEnumerate.add((BackedModel) item);
+                }
+                if (collectionTuple.getValue1()
+                        != items.length) { // and the size has changed since we last updated
+                    currentStoreDirtyCollections.put(entry.getKey(), collectionTuple.getValue0());
+                }
             }
+        }
+
+        // Enumerate nested backed models first since they may trigger the parent to be dirty
+        for (BackedModel nestedBackedModel : nestedBackedModelsToEnumerate) {
```

Comment:
We could do some error handling for cases where collections are not an instance of BackedModel

Reply:
If the collection doesn't contain `BackedModel`s, `nestedBackedModelsToEnumerate` will be empty:
https://github.com/microsoft/kiota-java/pull/1558/files#diff-d7e6365ce63b26ccd48ddba99ccc90260d769259ee24bd1987812fd72be950a8R245